### ISSUE TITLE
Add documentation

### ISSUE THREAD

**Issue by mbeutel** [gsl-lite/gsl-lite#198]
The Guidelines Support Library is sort-of documented in the [C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#S-gsl), so one could argue that GSL implementations such as *gsl-lite* only need to document their configuration switches.

However, *gsl-lite* offers a variety of additional features which are currently undocumented (e.g. the `gsl_HAVE_*` family of macros, language level-sensitive macros like `gsl_nullptr`, or `ExpectsAudit()`).

I think we should add dedicated documentation for all of *gsl-lite*'s features and some guidelines for proper use (cf. #194). 

My current plan is to use Sphinx for this, perhaps together with Doxygen and Breathe.

**Comment by petamas** [comment]
> However, gsl-lite offers a variety of additional features which are currently undocumented (e.g. the gsl_HAVE_* family of macros, language level-sensitive macros like gsl_nullptr, or ExpectsAudit()).

Also, the documentation in the Guidelines is not that exact, so not only the "additional" parts should be added to the documentation, but everything.

I don't know Sphinx, is it capable to handle the various #ifdefs we use in the lib? Like, how would the documentation of not_null's constructors look?

**Comment by mbeutel** [comment]
> Also, the documentation in the Guidelines is not that exact, so not only the "additional" parts should be added to the documentation, but everything.

Absolutely. We document our interpretation of the Guidelines w.r.t. the GSL, and whatever we provide in addition.


> I don't know Sphinx, is it capable to handle the various #ifdefs we use in the lib? Like, how would the documentation of not_null's constructors look?

Sphinx is just a documentation generator and doesn't have an understanding of C++; that is why it is often combined with Doxygen + Breathe. Doxygen itself is a great tool but I'm not a huge fan of the reference-style HTML output it produces. With Sphinx it is much easier to control the final result; you write plain Markdown in whatever structure you want, and you refer to Breathe if you want to insert a class declaration or function signature.

**Comment by mbeutel** [comment]
There should be a "why *gsl-lite*" section which should reiterate some of the points made here:
https://www.reddit.com/r/cpp/comments/ezqvuf/gsllite_v036_released/fgptxjo
